### PR TITLE
support publishing of ad hoc metrics

### PR DESCRIPTION
**Pull request by cluther**
Implements ZPS-4629.

### CODE REVIEW

File: ZenPacks/zenoss/PythonCollector/zenpython.py [hunk 1]
```diff
@@ -631,6 +641,26 @@ def storeValues(self, values):
...
                                 **WRITERRD_ARGS)
 
+    @inlineCallbacks
+    def publishMetrics(self, metrics):
+        if not (metrics and self.writeMetricWithMetadata):
+            returnValue(None)
+
+        self.state = PythonCollectionTask.STATE_PUBLISH_METRICS
+
+        writer = self._dataService.metricWriter()
```
Comment (joshw) [inline ZenPacks/zenoss/PythonCollector/zenpython.py]: Not really a comment on this change, but ZenCollector maintains a counter of metrics written, for the 'datapoints' graph on the collector..  I wonder if it should differentiate between the number sent to cloud vs locally.

File: README.html [hunk 2]
```diff
@@ -278,8 +278,8 @@ <h3 id="using-the-python-data-source-type-directly">Using the Python Data Source
         """
         Called only on success. After onResult, before onComplete.
 
-        You should return a data structure with zero or more events, values
-        and maps.
+        You should return a data structure with zero or more events, values,
+        metrics, and maps.
```
[the remark this answers is not in the document]
Comment (cluther) [reply README.html]: I agree. I'll add more documentation on that.

File: ZenPacks/zenoss/PythonCollector/datasources/PythonDataSource.py [hunk 3]
```diff
@@ -162,11 +166,20 @@ def new_data(self):
...
         return {
-            'values': defaultdict(dict),
+            'values': collections.defaultdict(dict),
+            'metrics': [],
             'events': [],
             'maps': [],
             }
 
+    @staticmethod
+    def get_metric(name, value, timestamp=None, tags=None):
```
[the remark this answers is not in the document]
Comment (cluther) [reply ZenPacks/zenoss/PythonCollector/datasources/PythonDataSource.py]: I agree that the naming is inconsistent. Do you recommend that I make this new_metric? Obviously we can't change new_data. metric does seem to be the defacto name for what this is.

File: ZenPacks/zenoss/PythonCollector/datasources/PythonDataSource.py [hunk 4]
```diff
@@ -193,9 +206,114 @@ def remote_myMethod(self, someParameter):
...
+        a time that is inconvenient to waiting for collect to be called.
+
+        This method is called by publishData() which can be used if there's
+        more than just events to publish.
+
+        Note: This method cannot be overridden. It is injected by zenpython.
+
+        """
+
+    def publishValues(self, values):
```
[the remark this answers is not in the document]
Comment (cluther) [reply ZenPacks/zenoss/PythonCollector/datasources/PythonDataSource.py]: values, events, and metrics are already buffered via the usual mechanisms. values and metrics are published to the local Redis and are only later consumed by the collector's MetricShipper to be shipped to the central MetricConsumer. Events are queued in-memory and flushed up to zenhub periodically. Even maps are "fired-and-forgotten" when running in cycle mode.

There's no such thing as immediate for any of them.

File: ZenPacks/zenoss/PythonCollector/datasources/PythonDataSource.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (joshw) [reply ZenPacks/zenoss/PythonCollector/datasources/PythonDataSource.py]: That seems fine to me

File: ZenPacks/zenoss/PythonCollector/datasources/PythonDataSource.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (joshw) [reply ZenPacks/zenoss/PythonCollector/datasources/PythonDataSource.py]: OK, then that's fine.

File: ZenPacks/zenoss/PythonCollector/zenpython.py [hunk 5]
```diff
@@ -631,6 +641,26 @@ def storeValues(self, values):
                                 timestamp=timestamp,
                                 **WRITERRD_ARGS)
 
+    @inlineCallbacks
+    def publishMetrics(self, metrics):
+        if not (metrics and self.writeMetricWithMetadata):
+            returnValue(None)
```
[the remark this answers is not in the document]
Comment (joshw) [reply ZenPacks/zenoss/PythonCollector/zenpython.py]: Understood- does that mean that metrics work on 5 and 6?   I was thinking that was cloud specific

[the remark this answers is not in the document]
Comment (cluther) [reply ZenPacks/zenoss/PythonCollector/zenpython.py]: I guess they'd "sort of work" for 5 and 6. There'd just be no easy way for the user to see them.

**Pull request merged**